build/src/lib.rs

```diff
@@ -40,50 +76,148 @@ impl Binary {
-        fs::write(path, &buffer)?;
+    Ok(Cursor::new(buffer))
+}
+
+fn construct_cache_path(
+    version: &str,
+    lang_code: &str,
+    binary: Binary,
+    cache_dir: Option<&PathBuf>,
+    transform_path_fn: Option<&Box<dyn TransformPathFn>>,
+) -> Result<Option<PathBuf>> {
+    let filename = binary.filename(lang_code);
+
+    cache_dir
+        .map(move |dir| {
+            let path = dir.join(version).join(lang_code).join(&filename);
+            Ok(if let Some(transform_path_fn) = transform_path_fn {
+                transform_path_fn(path)?
+            } else {
+                path
+            })
+        })
+        .transpose()
+}
+
+/// Returns a `dyn Read` type, that is either directly feed
+/// from an in-memory slice or from the on disk cache.
+/// If the on-disk cache is disabled or is not present,
+/// it will attempt to download it via [`obtain_binary_from_github_release`].
+/// The data is then written into `dest`.
+/// Also updates the cache.
+fn obtain_binary_cache_or_github(
+    version: &str,
+    lang_code: &str,
+    binary: Binary,
+    cache_dir: Option<&PathBuf>,
+    transform_path_fn: Option<&Box<dyn TransformPathFn>>,
+    transform_data_fn: Option<&Box<dyn TransformDataFn<Cursor<Vec<u8>>>>>,
+) -> Result<Cursor<Vec<u8>>> {
+    let cache_path = construct_cache_path(
+        version,
+        lang_code,
+        binary,
+        cache_dir.clone(),
+        transform_path_fn,
+    )?;
+
+    // if the file can be read, the data is already cached and the transform was applied before
+    if let Some(ref cache_path) = cache_path {
+        if let Ok(bytes) = fs::read(cache_path) {
+            if bytes.len() > 256 {
+                return Ok(Cursor::new(bytes));
+            }
+        }
     }
 
-    Ok(Cursor::new(buffer))
+    // the binencoded data from github
+    let reader_binenc = obtain_binary_from_github_release(version, lang_code, binary)?;
+
+    // apply the transform if any to an intermediate buffer
+    let mut reader_transformed = if let Some(transform_data_fn) = transform_data_fn {
+        // TODO this is not optimal, the additional copy is a bit annoying
+        let mut intermediate = Vec::<u8>::with_capacity(4096 * 1024);
+        transform_data_fn(reader_binenc, Cursor::new(&mut intermediate))?;
+        Cursor::new(intermediate)
+    } else {
+        reader_binenc
+    };
+
+    // update the cache entry
+    if let Some(ref cache_path) = cache_path {
+        fs::create_dir_all(cache_path.parent().expect("path must have parent"))?;
+        let mut cache_file = fs::OpenOptions::new()
+            .truncate(true)
+            .create(true)
+            .write(true)
+            .open(cache_path)?;
+        io::copy(&mut reader_transformed, &mut cache_file)?;
+    }
+
+    // move the cursor back to the beginning
+    reader_transformed.seek(SeekFrom::Start(0_u64))?;
+
+    Ok(reader_transformed)
+}
+
+fn assure_binary_availability(
```

Comment:
Maybe this should be called something else? `assure_*` sounds like something without any side effects to me but this function writes the binary to `out`.